core/src/main/java/com/sequenceiq/cloudbreak/controller/StackCreatorService.java

```diff
@@ -379,9 +383,66 @@ void fillInstanceMetadata(DetailedEnvironmentResponse environment, Stack stack)
+                        instanceGroup.getNotDeletedAndNotZombieInstanceMetaDataSet());
+                prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, instanceGroup);
+            }
+        }
+    }
+
+    @VisibleForTesting
+    public void datalakeMultiAZCalculator(Stack stack, Map<String, String> subnetAzPairs, String stackSubnetId, String stackAz) {
+        LOGGER.info("Setup multiAZ network for {}", stack);
+        for (InstanceGroup instanceGroup : sortInstanceGroups(stack)) {
+            if (!instanceGroup.getGroupName().equals(InstanceGroupName.AUXILIARY.getName())
+                && !instanceGroup.getGroupName().equals(InstanceGroupName.GATEWAY.getName())) {
+                LOGGER.info("calculating multiAZ to {}", instanceGroup.getGroupName());
+                multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs, instanceGroup.getInstanceGroupNetwork(),
+                        instanceGroup.getNotDeletedAndNotZombieInstanceMetaDataSet());
+                prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, instanceGroup);
+            }
+        }
+        Optional<InstanceGroup> auxiliaryHostGroup = getInstanceGroupByInstanceGroupName(stack, InstanceGroupName.AUXILIARY);
+        Optional<InstanceGroup> gatewayHostGroup = getInstanceGroupByInstanceGroupName(stack, InstanceGroupName.GATEWAY);
+        if (auxiliaryHostGroup.isEmpty() && gatewayHostGroup.isEmpty()) {
+            return;
+        }
+
+        Set<InstanceMetaData> mergedInstanceMetaData = new HashSet<>();
+        mergedInstanceMetaData.addAll(getInstanceMetadata(stack, InstanceGroupName.GATEWAY));
+        mergedInstanceMetaData.addAll(getInstanceMetadata(stack, InstanceGroupName.AUXILIARY));
+        AtomicReference<InstanceGroupNetwork> instanceGroupNetwork = new AtomicReference<>(new InstanceGroupNetwork());
+        auxiliaryHostGroup.ifPresentOrElse(hostgroup -> instanceGroupNetwork.set(auxiliaryHostGroup.get().getInstanceGroupNetwork()),
+                () -> instanceGroupNetwork.set(gatewayHostGroup.get().getInstanceGroupNetwork()));
+
+        multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs, instanceGroupNetwork.get(), mergedInstanceMetaData);
+        LOGGER.info("Auxiliary/Gateway hostgroup's multiAZ params: {}", mergedInstanceMetaData.iterator().next());
+        if (auxiliaryHostGroup.isPresent()) {
+            prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, auxiliaryHostGroup.get());
+        }
+        if (gatewayHostGroup.isPresent()) {
+            prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, gatewayHostGroup.get());
+        }
+    }
+
+    private Set<InstanceMetaData> getInstanceMetadata(Stack stack, InstanceGroupName groupName) {
+        AtomicReference<Set<InstanceMetaData>> instanceMetaDataSet = new AtomicReference<>(new HashSet<>());
```

Comment:
Why the `AtomicReference` is necessary here?

Reply:
instanceMetaDataSet is used in a lambda expression.